bdns/dns.go

```diff
@@ -283,29 +363,79 @@ func isPrivateV4(ip net.IP) bool {
 	return false
 }
 
-// LookupHost sends a DNS query to find all A records associated with the
-// provided hostname. This method assumes that the external resolver will chase
-// CNAME/DNAME aliases and return relevant A records.  It will retry requests in
-// the case of temporary network errors. It can return net package,
-// context.Canceled, and context.DeadlineExceeded errors.
-func (dnsResolver *DNSResolverImpl) LookupHost(ctx context.Context, hostname string) ([]net.IP, error) {
-	var addrs []net.IP
-	dnsType := dns.TypeA
-	r, err := dnsResolver.exchangeOne(ctx, hostname, dnsType, dnsResolver.aStats)
-	if err != nil {
-		return addrs, &DNSError{dnsType, hostname, err, -1}
+func isPrivateV6(ip net.IP) bool {
+	for _, net := range privateV6Networks {
+		if net.Contains(ip) {
+			return true
+		}
 	}
-	if r.Rcode != dns.RcodeSuccess {
-		return nil, &DNSError{dnsType, hostname, nil, r.Rcode}
+	return false
+}
+
+var errSkippedV6 = errors.New("skipped lookup of AAAA")
+
+// LookupHost sends a DNS query to find all A and AAAA records associated with
+// the provided hostname. This method assumes that the external resolver will
+// chase CNAME/DNAME aliases and return relevant records.  It will retry
+// requests in the case of temporary network errors. It can return net package,
+// context.Canceled, and context.DeadlineExceeded errors, all wrapped in the
+// DNSError type.
+func (dnsResolver *DNSResolverImpl) LookupHost(ctx context.Context, hostname string) ([]net.IP, error) {
+	var respA, respAAAA *dns.Msg
+	var errA, errAAAA error
+	var wg sync.WaitGroup
+
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		respA, errA = dnsResolver.exchangeOne(ctx, hostname, dns.TypeA, dnsResolver.aStats)
+		if errA != nil {
+			errA = &DNSError{dns.TypeA, hostname, errA, -1}
+			return
+		}
+		if respA.Rcode != dns.RcodeSuccess {
+			errA = &DNSError{dns.TypeA, hostname, nil, respA.Rcode}
+		}
+	}()
+	go func() {
+		defer wg.Done()
+		if !dnsResolver.lookupIPV6 {
+			respAAAA = new(dns.Msg)
+			errAAAA = errSkippedV6
+			return
+		}
+		respAAAA, errAAAA = dnsResolver.exchangeOne(ctx, hostname, dns.TypeAAAA, dnsResolver.aaaaStats)
+		if errAAAA != nil {
+			errAAAA = &DNSError{dns.TypeAAAA, hostname, errAAAA, -1}
+			return
+		}
+		if respAAAA.Rcode != dns.RcodeSuccess {
+			errAAAA = &DNSError{dns.TypeAAAA, hostname, nil, respAAAA.Rcode}
+		}
+	}()
+
+	wg.Wait()
+
+	if errA != nil && errAAAA != nil {
```

Comment:
I think amalgamating both errors is not likely to be informative. The most common case is going to be that both time out (or NXDOMAIN), in which case returning a single timeout error is sufficient.